server/lib/nl/nl_utterance.py

```diff
@@ -0,0 +1,226 @@
+# associated with a user issued query. The past utterances
+# form the context saved in the client and shipped to the server.
+@dataclass
+class Utterance:
+  # Linked list of past utterances
+  prev_utterance: Utterance
+  # Unmodified user-issued query
+  query: str
+  # Result of classification
+  detection: Detection
+  # A characterization of the query
+  query_type: ClassificationType
+  # Primary places
+  places: List[Place]
+  # Primary variables
+  svs: List[str]
+  # List of detected classifications
+  classifications: List[NLClassifier]
+  # Computed chart candidates.
+  chartCandidates: List[ChartSpec]
+  # Final ranked charts from which Chart Config proto is generated.
+  rankedCharts: List[ChartSpec]
+  # This is a list of places in the answer
+  # (e.g., top earthquake prone CA counties)
+  # TODO: Fill this up
+  answerPlaces: List[str]
+
+
+#
+# Helper functions for serializing/deserializing Utterance
+#
+
+
+def _save_places(places: List[Place]) -> List[Dict]:
+  places_dict = []
+  for p in places:
+    pdict = {}
+    pdict['dcid'] = p.dcid
+    pdict['name'] = p.name
+    pdict['place_type'] = p.place_type
+    places_dict.append(pdict)
+  return places_dict
+
+
+def _load_places(places_dict: List[Dict]) -> List[Place]:
+  places = []
+  for pdict in places_dict:
+    places.append(
+        Place(dcid=pdict['dcid'],
+              name=pdict['name'],
+              place_type=pdict['place_type']))
+  return places
+
+
+def _save_classifications(classifications: List[NLClassifier]) -> List[Dict]:
+  classifications_dict = []
+  for c in classifications:
+    cdict = {}
+    cdict['type'] = c.type
+
+    if isinstance(c.attributes, ContainedInClassificationAttributes):
+      cip = c.attributes.contained_in_place_type
+      if isinstance(cip, ContainedInPlaceType):
+        cdict['contained_in_place_type'] = cip.value
+      else:
+        cdict['contained_in_place_type'] = cip
+    elif isinstance(c.attributes, RankingClassificationAttributes):
+      cdict['ranking_type'] = c.attributes.ranking_type
+
+    classifications_dict.append(cdict)
+  return classifications_dict
+
+
+def _load_classifications(
+    classifications_dict: List[Dict]) -> List[NLClassifier]:
+  classifications = []
+  for cdict in classifications_dict:
+    attributes = SimpleClassificationAttributes()
+    if 'contained_in_place_type' in cdict:
+      attributes = ContainedInClassificationAttributes(
+          contained_in_place_type=ContainedInPlaceType(
+              cdict['contained_in_place_type']))
+    elif 'ranking_type' in cdict:
+      attributes = RankingClassificationAttributes(
+          ranking_type=[RankingType(r) for r in cdict['ranking_type']],
+          ranking_trigger_words=[])
+    classifications.append(
+        NLClassifier(type=ClassificationType(cdict['type']),
+                     attributes=attributes))
+  return classifications
+
+
+def _save_charts(charts: List[ChartSpec]) -> List[Dict]:
```

Comment:
instead of load/save (which is odd to see in a request handler) -- perhaps we can spell out the conversion types? something like chart_spec_to_dict